spike_multinvi.md

```diff
@@ -0,0 +1,205 @@
+private:
+    using tle_const_iterator = std::vector<sat355::TLE>::const_iterator;
+    using IteratorPairVector = std::vector<std::tuple<orbit_iterator, orbit_iterator>>;
+
+// Implementation
+private:
+    // SatOrbitMulti
+    virtual void OnCalculateOrbitalDataMulti(const tle_const_iterator& tleBegin, const tle_const_iterator& tleEnd, OrbitalDataVector& ioDataVector);
+    virtual void OnSortOrbitalVectorMulti(orbit_iterator& inBegin, orbit_iterator& inEnd);
+    virtual void OnSortMergeVector(orbit_iterator& ioBegin, orbit_iterator& ioMid, orbit_iterator& ioEnd);
+
+// Data Members
+private:
+    std::size_t mNumThreads{0};
+};
+
+
+// Multi thread class: Has-a SatOrbitSingle
+class SatOrbitMulti
+{
+// Interface
+public:
+    SatOrbitMulti(std::size_t inNumThreads = 1);
+    virtual ~SatOrbitMulti() = default;
+
+    std::vector<sat355::TLE> ReadFromFile(int argc, char* argv[]);
+    std::vector<OrbitalData> CalculateOrbitalData(const std::vector<sat355::TLE>& tleVector);
+    void SortOrbitalVector(std::vector<OrbitalData>& ioOrbitalVector);
+    std::vector<std::vector<OrbitalData>> CreateTrains(const std::vector<OrbitalData>& orbitalVector);
+    void PrintTrains(const std::vector<std::vector<OrbitalData>>& trainVector);
+
+// Types
+private:
+    using tle_const_iterator = std::vector<sat355::TLE>::const_iterator;
+    using OrbitalDataVector = std::tuple<std::mutex, std::vector<OrbitalData>>;
+    using orbit_iterator = std::vector<OrbitalData>::iterator;
+    using IteratorPairVector = std::vector<std::tuple<orbit_iterator, orbit_iterator>>;
+
+// Implementation
+private:
+    // SatOrbit
+    virtual std::vector<sat355::TLE> OnReadFromFile(int argc, char* argv[]);
+    virtual void OnCalculateOrbitalData(const std::vector<sat355::TLE>& inTLEVector, OrbitalDataVector& ioDataVector);
+    virtual void OnSortOrbitalVector(std::vector<OrbitalData>& ioOrbitalVector);
+    virtual std::vector<std::vector<OrbitalData>> OnCreateTrains(const std::vector<OrbitalData>& orbitalVector);
+    virtual void OnPrintTrains(const std::vector<std::vector<OrbitalData>>& trainVector);
+
+    // SatOrbitMulti
+    virtual void OnCalculateOrbitalDataMulti(const tle_const_iterator& tleBegin, const tle_const_iterator& tleEnd, OrbitalDataVector& ioDataVector);
+    virtual void OnSortOrbitalVectorMulti(orbit_iterator& inBegin, orbit_iterator& inEnd);
+    virtual void OnSortMergeVector(orbit_iterator& ioBegin, orbit_iterator& ioMid, orbit_iterator& ioEnd);
+
+// Data Members
+private:
+    std::size_t mNumThreads{0};
+    SatOrbitSingle mSatOrbit{};
+};
+```
+
+- Has-a
+Pros:
+Increased modularity
+Simpler to implement
+No need to change API if dependencies update
+Aggregation preferred
+
+Cons:
+Users of the old API must update their code to utilize new APIs
+No backwards compatibility
+
+
+- Is-a
+Pros:
+Follows conventions of NVI
+Inheritance allows for multiple implementations
+
+Cons:
+More complex
+Inheritance not preferred
+Implementation not visible
```

Comment:
what does "not visible" mean? does that mean the implementation is somehow visible with "has-a"?